cpu/arm7_common/VIC.c

```diff
@@ -14,20 +14,20 @@
 static inline unsigned __get_cpsr(void)
 {
     unsigned long retval;
-    __asm__ volatile(" mrs  %0, cpsr" : "=r"(retval) : /* no inputs */);
+    __asm__ volatile(" mrs  %0, cpsr" : "=r"(retval) : /* no inputs */ : "memory");
     return retval;
 }
 
 int irq_is_in(void)
 {
     int retval;
-    __asm__ volatile(" mrs  %0, cpsr" : "=r"(retval) : /* no inputs */);
+    __asm__ volatile(" mrs  %0, cpsr" : "=r"(retval) : /* no inputs */ : "memory");
```

Comment:
why here?

Reply:
I was thinking about control dependencies and speculative stores.

E.g.

```C
if (irq_is_in()) {
    /* no lock required */
    x = 3;
} 
else {
    lock();
    x = 7;
    unlock();
}
```

could be transformed into:

```C
x = 3;
if (!irq_is_in()) {
    lock();
    x = 7;
    unlock();
}
```

That unprotected `x = 3` outside of an ISR could bring a share data structure into an corrupted state...

Reply:
Hmm, not a good example. `lock()`/`unlock()` would need to be `irq_disable()`/`irq_restore()` for that to be correct. But then it would be more elegant to call those regardsless of interrupt-/thread-context. But it would be technically correct, even though inelegant.

Reply:
> could be transformed into:

Uff, is the compiler allowed to do this? Do you have a reference?

For local variables, that transformation wouldn't matter. This would only affect accesses to global variables (or writes through pointers).

Reply:
> Uff, is the compiler allowed to do this?

If a transformation does not change visible behavior from the compilers point of view, it is legal. The compiler is unaware of concurrent memory accesses. So speculative stores to `x` in the example above would be totally valid from the compilers viewpoint (even if `x` is a global), as the compiler does not take into account that other threads could read `x` at that very instance when the speculative store was done and not fixed subsequently.

> Do you have a reference?

This article. https://lwn.net/Articles/586838/ (If you want to fast-forward to relevant part search for speculative stores

In more detail: If the inline assembly in `irq_is_in()` is only marked as `volatile`, but no (compiler) barrier is added. Thus, non-`volatile` accesses can be reordered with regard to `irq_is_in()`. The store `x = 3` is neither `volatile` nor guarded by a memory barrier, so it could be moved up. If that store was done incorrectly, the subsequent `x = 7` will fix it. So in a single-threaded context this transformation will work just fine - which is all the compiler needs to prove.

Reply:
ok